ms2query/ms2library.py

```diff
@@ -0,0 +1,332 @@
+        is stored. Which column names/info is stored can be found in
+        collect_data_for_tanimoto_prediction_model.
+
+        Args:
+        ------
+        query_spectra:
+            The spectra for which info about matches should be collected
+        progress_bar:
+            If true a progress bar is shown. Default is False
+        """
+        # Gets a preselection of spectra for all query_spectra
+        dict_with_preselected_spectra = self.pre_select_spectra(query_spectra)
+
+        # Run neural network model over all found spectra and add the predicted
+        # scores to a dict with the spectra in dataframes
+        dict_with_preselected_spectra_info = {}
+        for query_spectrum in tqdm(query_spectra,
+                                   desc="collecting matches info",
+                                   disable=not progress_bar):
+            spectrum_id = query_spectrum.get(self.spectrum_id_column_name)
+            matches_with_info = \
+                self._collect_data_for_ms2query_model(
+                    query_spectrum,
+                    dict_with_preselected_spectra[spectrum_id])
+            dict_with_preselected_spectra_info[spectrum_id] = matches_with_info
+        return dict_with_preselected_spectra_info
+
+    def pre_select_spectra(self,
+                           query_spectra: List[Spectrum],
+                           nr_of_spectra: int = 20
+                           ) -> Dict[str, List[str]]:
+        """Returns dict with spectrum IDs of spectra with highest ms2ds score
+
+        The keys are the query spectrum_ids and the values a list of the
+        preselected spectrum_ids for each query spectrum.
+
+        Args:
+        ------
+        query_spectra:
+            spectra for which a preselection of possible library matches should
+            be done
+        nr_of_spectra:
+            How many top spectra should be selected. Default = 20
+        """
+        dict_with_preselected_spectra = {}
+        # Select top nr of spectra
+        for query_spectrum in query_spectra:
+            query_spectrum_id = query_spectrum.get(
+                self.spectrum_id_column_name)
+            ms2ds_scores = self._get_all_ms2ds_scores(query_spectrum)
+            ms2ds_scores_np = ms2ds_scores["ms2ds_score"].to_numpy()
+            # Get the indexes of the spectra with the highest ms2ds scores
+            indexes_of_top_spectra = np.argpartition(
+                ms2ds_scores_np,
+                -nr_of_spectra,
+                axis=0)[-nr_of_spectra:]
+            # Select the spectra with the highest score
+            selected_spectra = list(ms2ds_scores["ms2ds_score"].iloc[
+                indexes_of_top_spectra].index)
+            # Store selected spectra in dict
+            dict_with_preselected_spectra[query_spectrum_id] = selected_spectra
+        return dict_with_preselected_spectra
+
+    def _get_all_ms2ds_scores(self, query_spectrum: Spectrum) -> pd.DataFrame:
```

Comment:
Can be dealt with separately (--> new issue to keep track):
The score computation will probably be even faster if the full calculation is done at once between ALL query spectra and all library spectra (because the actual cosine function is numba compiled). So maybe we could switch from `query_embedding` to `query_embeddings` ?

Reply:
Good to know, I changed it right away